common/string_helpers.cpp

```diff
@@ -81,8 +82,40 @@ auto UnescapeStringLiteral(llvm::StringRef source, const int hashtag_num,
           ret.push_back(16 * *c1 + *c2);
           break;
         }
-        case 'u':
-          CARBON_FATAL() << "\\u is not yet supported in string literals";
+        case 'u': {
+          i++;
+          if (i >= source.size() || source[i] != '{')
+            return std::nullopt;
+          unsigned int unicode_int = 0;
+          i++;
+          while (i < source.size() && source[i] != '}') {
+            char hex_val;
+            if (source[i] >= '0' && source[i] <= '9') {
+              hex_val = source[i] - '0';
+            } else if (source[i] >= 'A' && source[i] <= 'F') {
+              hex_val = source[i] - 'A' + 10;
+            } else {
+              return std::nullopt;
+            }
+            unicode_int = unicode_int << 4;
+            unicode_int += hex_val;
+            i++;
+          }
+          if (i >= source.size())
+            return std::nullopt;
+          char temp[4];
+          char *result_two = &temp[0];
+          char *result = &temp[0];
```

Comment:
As mentioned previously, I am not using it right now since it threw an unusual error. But I will check back on this.

Reply:
@jonmeow, thanks for the suggestion. Actually, I was wondering if you could help me understand something with that part of the code. I didn't originally want to make all of these extra pointers. The plan was to somehow pass a pointer to the temp array straight into ConvertCodePointToUTF8. However, I couldn't get it to work, mainly because I didn't properly understand the type being used, which was a char *&. 

From what I understand, that's a reference to a char pointer, so I thought I could just pass in &temp, since temp is a char * (an array, which is really a char pointer). However, that didn't work, and after trying a few more things I eventually got this, which works, but feels unnecessarily messy. Is there a better way to do this?

Reply:
@vinsdragonis What was the error?

@Pritjam Sorry, I failed to include my `result.append` suggestion (added below). You can't just pass in `&temp` because ConvertCodePointToUTF8 requires a reference parameter to indicate the end of the UTF8 sequence: [documentation](https://llvm.org/doxygen/namespacellvm.html#ad43161495477e1f7f4ec404e2bb9b075)

Reply:
I managed to get this working, though I am not exactly sure this is how it is meant to be. After eliminating the previously mentioned while loop, I got rid of `char *result = &temp[0];` as it was not used anywhere else. That seemed to fix the problem for me.